pkg/config/factory.go

```diff
@@ -0,0 +1,881 @@
+	Configs []*NamespacedName
+	Targets []*ClusterTarget
+}
+
+// Factory handle the config
+type Factory interface {
+	ParseTemplate(defaultName string, content []byte) (*Template, error)
+	ParseConfig(ctx context.Context, template NamespacedName, meta Metadata) (*Config, error)
+
+	LoadTemplate(ctx context.Context, name, ns string) (*Template, error)
+	CreateOrUpdateConfigTemplate(ctx context.Context, ns string, it *Template) error
+	DeleteTemplate(ctx context.Context, ns, name string) error
+	ListTemplates(ctx context.Context, ns, scope string) ([]*Template, error)
+
+	ReadConfig(ctx context.Context, namespace, name string) (map[string]interface{}, error)
+	GetConfig(ctx context.Context, namespace, name string, withStatus bool) (*Config, error)
+	ListConfigs(ctx context.Context, namespace, template, scope string, withStatus bool) ([]*Config, error)
+	DeleteConfig(ctx context.Context, namespace, name string) error
+	CreateOrUpdateConfig(ctx context.Context, i *Config, ns string) error
+
+	CreateOrUpdateDistribution(ctx context.Context, ns, name string, ads *CreateDistributionSpec) error
+	ListDistributions(ctx context.Context, ns string) ([]*Distribution, error)
+	DeleteDistribution(ctx context.Context, ns, name string) error
+	MergeDistributionStatus(ctx context.Context, config *Config, namespace string) error
+}
+
+// NewConfigFactory create a config factory instance
+func NewConfigFactory(cli client.Client) Factory {
+	return &kubeConfigFactory{cli: cli, apiApply: apply.NewAPIApplicator(cli)}
+}
+
+type kubeConfigFactory struct {
+	cli      client.Client
+	apiApply *apply.APIApplicator
+}
+
+// ParseTemplate parse a config template instance form the cue script
+func (k *kubeConfigFactory) ParseTemplate(defaultName string, content []byte) (*Template, error) {
+	cueScript := script.BuildCUEScriptWithDefaultContext(icontext.DefaultContext, content)
+
+	value, err := cueScript.ParseToValue(false)
+	if err != nil {
+		return nil, fmt.Errorf("the cue script is invalid:%w", err)
+	}
+	name, err := value.GetString("metadata", "name")
+	if err != nil {
+		if defaultName == "" {
+			return nil, fmt.Errorf("fail to get the name from the template metadata: %w", err)
+		}
+	}
+	if defaultName != "" {
+		name = defaultName
+	}
+	schema, err := cueScript.ParsePropertiesToSchema()
+	if err != nil {
+		return nil, fmt.Errorf("the properties of the cue script is invalid:%w", err)
+	}
+	alias, err := value.GetString("metadata", "alias")
+	if err != nil && !IsFieldNotExist(err) {
+		klog.Warningf("fail to get the alias from the template metadata: %s", err.Error())
+	}
+	scope, err := value.GetString("metadata", "scope")
+	if err != nil && !IsFieldNotExist(err) {
+		klog.Warningf("fail to get the scope from the template metadata: %s", err.Error())
+	}
+	sensitive, err := value.GetBool("metadata", "sensitive")
+	if err != nil && !IsFieldNotExist(err) {
+		klog.Warningf("fail to get the scope from the template metadata: %s", err.Error())
```

Comment:
does it necessary data?

Reply:
The alias, scope, and sensitive are not required fields. So here ignore the not exist error, and print the other errors.